enhancements/aiops/20230203-introduce-streaming-dataloader-in-log-anomaly-detection-model-training.md

```diff
@@ -0,0 +1,40 @@
+# Title: 
+implement streaming data-loader for log anomaly detection model training
+
+## Summary: 
+ AIOps' log anomaly detection model can train model using a streaming data-loader, such that it doesn't have to load all training dataset into memory, instead, it only dumps a subset of training dataset each time and trains with that subset of data. This feature significantly reduces the memory resource pressure in the model training service.
+
+## Use case: 
+when Opni's user launches a workload log anomaly detection training job with huge size of training data (for example, 500Gb logs), the current approach that loads all data into memory won't be feasible. The streaming data-loader can resolve this problem because it only loads a reasonable subset of training dataset each time.
+
+## Benefits: 
+* resolve the memory pressure issue in the scenario that training data size is large.
+
+## Impact: 
+* This feature will prevent the model training service from running out of memory.
+
+## Implementation details: 
+* the [method that queries training data](https://github.com/rancher/opni-inference-service/blob/main/opnilog-inference-service/opnilog_trainer.py#L46) from Opensearch will return a generator, so it's iterable and it won't keep all the data in memory. It yields each scroll of the search results.
+* To implement a streaming data-loader similar to [`IterableDataset`](https://pytorch.org/docs/stable/data.html#torch.utils.data.IterableDataset) in this [script](https://github.com/rancher/opni-inference-service/blob/main/models/opnilog/opnilog_parser.py). More specifically, an `__iter__()` function will be implemented in this data-loader, which returns an iterator if training dataset.
+* the training method [`train()`](https://github.com/rancher/opni-inference-service/blob/main/models/opnilog/opnilog_parser.py#L75) will be slightly adjusted to the new streaming data-loader.
```

Comment:
Will you be using Mult-process loading? If so, how many workers and why? Include those details in this section

Reply:
updated.